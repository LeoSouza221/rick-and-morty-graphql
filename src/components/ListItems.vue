<template>
  <v-container>
    <v-data-iterator
      :items="orderedResults"
      :page.sync="page"
      :items-per-page="itemsPerPage"
      :loading="loading"
      hide-default-footer
    >
      <template v-slot:loading>
        <v-progress-linear
          indeterminate
          color="primary"
        ></v-progress-linear>
      </template>
      <template v-slot:header>
        <v-row class="mt-5">
          <v-col cols="12" sm="3">
            <div class="d-flex align-center">
              <h3 class="title">{{ listName }}</h3>
            </div>
          </v-col>
          <v-spacer></v-spacer>
          <v-col cols="12" sm="6">
            <v-row>
              <v-col cols="12" sm="6">
                <v-select
                  v-model="sortBy"
                  flat
                  dense
                  hide-details
                  :items="keys"
                  item-text="name"
                  item-value="value"
                  prepend-inner-icon="mdi-sort-descending"
                  label="Sort by"
                  class="mb-1"
                ></v-select>
              </v-col>
              <v-col cols="12" sm="6">
                <v-text-field
                  v-model="searchValue"
                  clearable
                  dense
                  flat
                  hide-details
                  prepend-inner-icon="mdi-magnify"
                  label="Search"
                  class="search-input"
                  @input="searchText"
                ></v-text-field>
              </v-col>
            </v-row>
          </v-col>
        </v-row>
        <v-divider></v-divider>
      </template>
      <template v-slot:default="props">
        <v-row class="mt-5">
          <slot name="info-cards" v-bind:items="props.items">
          </slot>
        </v-row>
      </template>
    </v-data-iterator>
    <div class="text-center pt-2">
      <v-pagination
        v-model="page"
        circle
        :length="pageCount"
        :total-visible="5"
        color="warning"
        @input="changePage"
      ></v-pagination>
    </div>
  </v-container>
</template>

<script>
export default {
  name: 'ListItems',

  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    pageCount: {
      type: Number,
      required: true,
    },
    listName: {
      type: String,
      default: 'List',
    },
    newSearch: {
      type: Function,
    },
    results: {
      type: Array,
      required: true,
    },
    keys: {
      type: Array,
      required: true,
    },
  },

  data: () => ({
    timeout: null,
    itemsPerPage: 20,
    page: 1,
    sortBy: '',
    searchValue: '',
  }),

  computed: {
    orderedResults() {
      const newArray = this.results;
      return newArray.sort((previous, currentb) => {
        if (previous[this.sortBy] < currentb[this.sortBy]) {
          return -1;
        }
        if (previous[this.sortBy] > currentb[this.sortBy]) {
          return 1;
        }
        return 0;
      });
    },
  },

  methods: {
    doNewSearch() {
      const search = {
        page: this.page,
        name: this.searchValue,
      };
      this.$emit('newSearch', search);
    },

    searchText(text) {
      clearTimeout(this.timeout);

      this.timeout = setTimeout(() => {
        console.log('aqui', text);
        if (typeof text === 'object' || (text.length && text.length >= 3)) {
          this.page = 1;
          this.doNewSearch();
        }
      }, 1000);
    },

    changePage() {
      this.doNewSearch();
    },
  },
};
</script>
