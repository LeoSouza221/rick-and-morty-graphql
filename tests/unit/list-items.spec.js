import ListItems from '../../src/components/ListItems.vue';
import Vue from 'vue';
import Vuetify from 'vuetify';
import {createLocalVue, mount} from '@vue/test-utils';

describe('ListItems.vue', () => {
  let localVue;
  let vuetify;
  let wrapper;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    wrapper = mount(ListItems, {
      localVue,
      vuetify,
      propsData: {
        listName: 'Test Title',
        results: [
          {
            text: 'A',
            number: 2,
          },
        ],
        pageCount: 1,
        keys: [],
      },
      slots: {
        'info-cards': `<div class="card-slot">Card Slot</div>`
      }
    })
  });

  it('should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match title', () => {
    const title = wrapper.find('.title');

    expect(title.text()).toBe('Test Title');
  });

  it('should find slot', () => {
    const cardSlot = wrapper.find('.card-slot');

    expect(cardSlot.text()).toBe('Card Slot');
  });

  it('should pagination match with number of prop page count', async () => {
    let paginationItems = wrapper.findAll('button.v-pagination__item');
    const newPage = 2;

    expect(paginationItems.length).toBe(wrapper.vm.pageCount);

    await wrapper.setProps({ pageCount: newPage })

    paginationItems = wrapper.findAll('button.v-pagination__item');

    expect(paginationItems.length).toBe(newPage);
  });

  it('should pagination emit event when change', async () => {
    const event = jest.fn()
    const newPage = 2;

    wrapper.vm.$on('newSearch', event)
    await wrapper.setProps({ pageCount: newPage })

    const paginationItems = wrapper.findAll('button.v-pagination__item');

    expect(event).toHaveBeenCalledTimes(0)
    expect(paginationItems.length).toBe(newPage);

    const pageTwo = paginationItems.at(1)
    pageTwo.trigger('click')

    expect(event).toHaveBeenCalledTimes(1)
  });

  it('should return ordered array', async () => {
    const resultsArray = [
      {
        number: 2,
      },
      {
        number: 1,
      },
    ];
    const sortBy = 'number';

    await wrapper.setData({ sortBy });
    await wrapper.setProps({ results: resultsArray });

    const [itemOne, itemTwo] = wrapper.vm.orderedResults
    expect(itemOne.number).toBe(1);
    expect(itemTwo.number).toBe(2);
  });

  it('should emit event when type more then 3 words in input', async () => {
    jest.useFakeTimers()
    const event = jest.fn()
    const textInput = wrapper.find('.search-input input[type="text"]')

    wrapper.vm.$on('newSearch', event)
    await textInput.setValue('som')

    expect(wrapper.vm.searchValue).toBe('som')
    expect(event).toHaveBeenCalledTimes(0)

    await textInput.setValue('some value')

    await Vue.nextTick()
    
    jest.advanceTimersByTime(1000);
    expect(event).toHaveBeenCalledTimes(1);
  });
});
