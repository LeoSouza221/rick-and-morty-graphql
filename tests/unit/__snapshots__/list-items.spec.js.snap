// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItems.vue should match with the snapshot 1`] = `
<div class="container">
  <div class="v-data-iterator">
    <div class="row mt-5">
      <div class="col-sm-3 col-12">
        <div class="d-flex align-center">
          <h3 class="title">Test Title</h3>
        </div>
      </div>
      <div class="spacer"></div>
      <div class="col-sm-6 col-12">
        <div class="row">
          <div class="col-sm-6 col-12">
            <div class="v-input mb-1 v-input--hide-details v-input--dense theme--light v-text-field v-text-field--solo-flat v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-4" class="v-input__slot">
                  <div class="v-input__prepend-inner">
                    <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-sort-descending theme--light"></i></div>
                  </div>
                  <div class="v-select__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">Sort by</label>
                    <div class="v-select__selections"><input id="input-4" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-sm-6 col-12">
            <div class="v-input search-input v-input--hide-details v-input--dense theme--light v-text-field v-text-field--solo-flat">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input__prepend-inner">
                    <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></div>
                  </div>
                  <div class="v-text-field__slot"><label for="input-8" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-8" type="text"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"></button></div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="row mt-5">
      <div class="card-slot">Card Slot</div>
    </div>
  </div>
  <div class="text-center pt-2">
    <nav role="navigation" aria-label="Pagination Navigation">
      <ul class="v-pagination v-pagination--circle theme--light">
        <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
        <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active warning">1</button></li>
        <li><button disabled="disabled" type="button" aria-label="Next page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
      </ul>
    </nav>
  </div>
</div>
`;
